### PR TITLE
Core: Add writer for unordered position deletes

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds a position delete writer that can handle unordered position deletes. This writer should allow us to avoid a local sort for **_some_** MERGE operations. Specifically, consider MERGE operations where 90% of data are inserts and the table is partitioned but no sort order is defined. Right now, we always request a local sort to order deletes. However, that sort can be useless for inserts if no sort order is defined and fanout writer is enabled. Moreover, ordering inserts may lead to a spill, which is expensive for wide tables and large tasks.

### CODE REVIEW

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java [hunk 1]
```diff
@@ -363,6 +389,60 @@ public void writeUnpartitionedClusteredPositionDeleteWriter(Blackhole blackhole)
...
+        closeableWriter.write(positionDelete, unpartitionedSpec, null);
+      }
+    }
+
+    blackhole.consume(writer);
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void writeUnpartitionedFanoutPositionDeleteWriterShuffled(Blackhole blackhole)
```
Comment (aokolnychyi) [inline spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java:420]: We should expect 5-15% overhead for the new buffering writer, which can still be beneficial for the job if we skip local ordering for inserts and potentially avoid spilling. This benchmark also does not take into account the cost to order records, it only tests the write performance. We will use this writer only if fanout is enabled. We should also explore Puffin delete files that would persist bitmaps directly.

```
Benchmark                                                                                                      Mode  Cnt           Score            Error   Units
ParquetWritersBenchmark.writeUnpartitionedClusteredPositionDeleteWriter                                          ss    5           6.004 ±          0.185    s/op
ParquetWritersBenchmark.writeUnpartitionedFanoutPositionDeleteWriter                                             ss    5           6.503 ±          0.171    s/op
ParquetWritersBenchmark.writeUnpartitionedFanoutPositionDeleteWriterShuffled                                     ss    5           6.616 ±          0.204    s/op
```

**Comment by aokolnychyi** [comment]
cc @singhpk234 @amogh-jahagirdar @RussellSpitzer @szehon-ho @flyrain @rdblue

**Review by singhpk234 — APPROVED** [review]
LGTM, Thanks @aokolnychyi !

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java (hunk 1 above)
Comment (singhpk234) [reply spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java:420]: > We should also explore Puffin delete files that would persist bitmaps directly

+1

File: core/src/main/java/org/apache/iceberg/deletes/PositionDelete.java [hunk 2]
```diff
@@ -38,6 +38,13 @@ public PositionDelete<R> set(CharSequence newPath, long newPos, R newRow) {
     return this;
   }
 
+  public PositionDelete<R> set(CharSequence newPath, long newPos) {
+    this.path = newPath;
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/deletes/PositionDelete.java:42]: Nit question: is it cleaner to have this constructor delegate to the other one?

File: core/src/main/java/org/apache/iceberg/deletes/SortingPositionOnlyDeleteWriter.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import org.apache.iceberg.io.FileWriter;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.util.CharSequenceWrapper;
+import org.roaringbitmap.longlong.PeekableLongIterator;
+import org.roaringbitmap.longlong.Roaring64Bitmap;
+
+/**
+ * A position delete writer that is capable of handling unordered deletes without rows.
```
Comment (szehon-ho) [inline core/src/main/java/org/apache/iceberg/deletes/SortingPositionOnlyDeleteWriter.java:34]: Nit: can we add javadoc to the PositionDeleteWriter, when we get a chance?

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java (hunk 1 above)
Comment (szehon-ho) [reply spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java:420]: About memory overhead (not sure any thing measures it now):  Should be just additional space of map (data_file_path => bitmaps)?  Will there be cases, esp in fanout, where a writer writes many delete of many data files, that will start to stress it?

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java [hunk 4]
```diff
@@ -93,14 +100,33 @@ public void setupBenchmark() {
             row -> transform.bind(Types.IntegerType.get()).apply(row.getInt(1))));
     this.rows = data;
 
-    this.positionDeleteRows =
-        RandomData.generateSpark(DeleteSchemaUtil.pathPosSchema(), NUM_ROWS, 0L);
+    this.positionDeleteRows = generatePositionDeletes(false /* shuffle */);
```
Comment (szehon-ho) [inline spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java:103]: Don't shuffle?

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java [hunk 5]
```diff
@@ -93,14 +100,33 @@ public void setupBenchmark() {
...
     this.partitionedSpec = table().specs().get(1);
   }
 
+  private Iterable<InternalRow> generatePositionDeletes(boolean shuffle) {
+    int numDeletes = NUM_DATA_FILES_PER_POSITION_DELETE_FILE * NUM_DELETED_POSITIONS_PER_DATA_FILE;
+    List<InternalRow> deletes = Lists.newArrayListWithExpectedSize(numDeletes);
+
+    for (int pathIndex = 0; pathIndex < NUM_DATA_FILES_PER_POSITION_DELETE_FILE; pathIndex++) {
+      UTF8String path = UTF8String.fromString("path/to/position/delete/file/" + UUID.randomUUID());
+      int step = 10;
```
Comment (szehon-ho) [inline spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java:117]: why not just make this outside?

**Review by szehon-ho — COMMENTED** [review]
Looks good to me too, left few comments

File: spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java (hunk 1 above)
Comment (aokolnychyi) [reply spark/v3.4/spark/src/jmh/java/org/apache/iceberg/spark/source/WritersBenchmark.java:420]: I ran this benchmark (100 data files, 50k deletes each, 5 million deletes total) with a GC profiler and did not see anything bad. Issues will arise when there are lots of unique data files. That's unlikely as we distribute by partition and this writer will still be disabled by default, so users will have to opt in explicitly.

File: core/src/main/java/org/apache/iceberg/deletes/SortingPositionOnlyDeleteWriter.java (hunk 3 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/deletes/SortingPositionOnlyDeleteWriter.java:34]: Will add in this PR.

**Comment by aokolnychyi** [comment]
Thanks, @singhpk234 @szehon-ho!

**Pull request merged**